BUG: upcasting of indices dtype from DIA to COO/CSR/BSR arrays

### Describe your issue.

While I understand the decision of not using downcasting for indices (cf. https://github.com/scipy/scipy/pull/18509), it seems that the conversion from DIA to COO/CSR/BSR matrices will make an upcast from `np.int32` to `np.int64` while it should not be needed. I assume that using the `dtype` of the `offsets` should be enough to be used during the conversion.

### Reproducing Code Example

```python
# %%
import numpy as np
from scipy import sparse

data = np.array([[1, 2, 3, 4]]).repeat(3, axis=0)
offsets = np.array([0, -1, 2], dtype=np.int32)
dia = sparse.dia_array((data, offsets), shape=(4, 4))
dia.__dict__

# %%
coo = dia.tocoo()
assert coo.col.dtype == np.int32, f"Wrong dtype: {coo.col.dtype}"

# %%
csr = dia.tocsr()
assert csr.indices.dtype == np.int32, f"Wrong dtype: {csr.indices.dtype}"

# %%
csc = dia.tocsc()
assert csc.indices.dtype == np.int32, f"Wrong dtype: {csc.indices.dtype}"
```

### Error message

```shell
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
[/Users/glemaitre/Documents/scratch/tmp.py](https://file+.vscode-resource.vscode-cdn.net/Users/glemaitre/Documents/scratch/tmp.py) in line 3
      [10](file:///Users/glemaitre/Documents/scratch/tmp.py?line=9) # %%
      [11](file:///Users/glemaitre/Documents/scratch/tmp.py?line=10) coo = dia.tocoo()
----> [12](file:///Users/glemaitre/Documents/scratch/tmp.py?line=11) assert coo.col.dtype == np.int32, f"Wrong dtype: {coo.col.dtype}"

AssertionError: Wrong dtype: int64

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
[/Users/glemaitre/Documents/scratch/tmp.py](https://file+.vscode-resource.vscode-cdn.net/Users/glemaitre/Documents/scratch/tmp.py) in line 3
      [14](file:///Users/glemaitre/Documents/scratch/tmp.py?line=13) # %%
      [15](file:///Users/glemaitre/Documents/scratch/tmp.py?line=14) csr = dia.tocsr()
----> [16](file:///Users/glemaitre/Documents/scratch/tmp.py?line=15) assert csr.indices.dtype == np.int32, f"Wrong dtype: {csr.indices.dtype}"

AssertionError: Wrong dtype: int64
```


### SciPy/NumPy/Python version and system information

```shell
SciPy version: '1.11.2'
```

--- a/scipy/sparse/_dia.py
+++ b/scipy/sparse/_dia.py
@@ -87,7 +87,7 @@ class _dia_base(_data_matrix):
             raise ValueError('offset array contains duplicate values')
 
     def __repr__(self):
-        format = _formats[self.getformat()][1]
+        format = _formats[self.format][1]
         return "<%dx%d sparse matrix of type '%s'\n" \
                "\twith %d stored elements (%d diagonals) in %s format>" % \
                (self.shape + (self.dtype.type, self.nnz, self.data.shape[0],
@@ -332,6 +332,11 @@ class _dia_base(_data_matrix):
         mask &= (self.data != 0)
         row = row[mask]
         col = np.tile(offset_inds, num_offsets)[mask.ravel()]
+        idx_dtype = self._get_index_dtype(
+            arrays=(self.offsets,), maxval=max(self.shape)
+        )
+        row = row.astype(idx_dtype, copy=False)
+        col = col.astype(idx_dtype, copy=False)
         data = self.data[mask]
         # Note: this cannot set has_canonical_format=True, because despite the
         # lack of duplicates, we do not generate sorted indices.

--- a/scipy/sparse/tests/test_base.py
+++ b/scipy/sparse/tests/test_base.py
@@ -4928,6 +4928,7 @@ def cases_64bit():
         'test_large_dimensions_reshape': 'test actually requires 64-bit to work',
         'test_constructor_smallcol': 'test verifies int32 indexes',
         'test_constructor_largecol': 'test verifies int64 indexes',
+        'test_tocoo_tocsr_tocsc_gh19245': 'test verifies int32 indexes',
     }
 
     for cls in TEST_CLASSES:

--- a/scipy/sparse/tests/test_base.py
+++ b/scipy/sparse/tests/test_base.py
@@ -4462,6 +4462,19 @@ class TestDIA(sparse_test_class(getset=False, slicing=False, slicing_assign=Fals
         inds_are_sorted = np.all(np.diff(flat_inds) > 0)
         assert m.has_canonical_format == inds_are_sorted
 
+    def test_tocoo_tocsr_tocsc_gh19245(self):
+        # test index_dtype with tocoo, tocsr, tocsc
+        data = np.array([[1, 2, 3, 4]]).repeat(3, axis=0)
+        offsets = np.array([0, -1, 2], dtype=np.int32)
+        dia = sparse.dia_array((data, offsets), shape=(4, 4))
+
+        coo = dia.tocoo()
+        assert coo.col.dtype == np.int32
+        csr = dia.tocsr()
+        assert csr.indices.dtype == np.int32
+        csc = dia.tocsc()
+        assert csc.indices.dtype == np.int32
+
 
 TestDIA.init_class()
 

